tests/check_pip

```diff
@@ -0,0 +1,193 @@
+
+# Set pip-licenses options
+LICENSE_OPTIONS="--from=mixed"
+
+function cleanup() {
+	rm -f ${LOCK_FILE} 2>/dev/null || : ; wait ;
+	# unset when done
+	unset LOCK_FILE 2>/dev/null || : ;
+	hash -d shlock 2>/dev/null || : ;
+}
+
+if [[ ( $(shlock -f ${LOCK_FILE} -p $$ ) -eq 0 ) ]] ; then
+		EXIT_CODE=0
+		trap 'cleanup 2>/dev/null || rm -f ${LOCK_FILE} 2>/dev/null || true ; wait ; exit 129 ;' SIGHUP || EXIT_CODE=129
+		trap 'cleanup 2>/dev/null || rm -f ${LOCK_FILE} 2>/dev/null || true ; wait ; exit 143 ;' SIGTERM || EXIT_CODE=143
+		trap 'cleanup 2>/dev/null || rm -f ${LOCK_FILE} 2>/dev/null || true ; wait ; exit 131 ;' SIGQUIT || EXIT_CODE=131
+		# SC2173 - https://github.com/koalaman/shellcheck/wiki/SC2173
+		#trap 'cleanup 2>/dev/null || rm -f ${LOCK_FILE} 2>/dev/null || true ; wait ; exit 1 ;' SIGSTOP || EXIT_CODE=7
+		trap 'cleanup 2>/dev/null || rm -f ${LOCK_FILE} 2>/dev/null || true ; wait ; exit 130 ;' SIGINT || EXIT_CODE=130
+		trap 'cleanup 2>/dev/null || rm -f ${LOCK_FILE} 2>/dev/null || true || true ; wait ; exit 137 ;' SIGABRT || EXIT_CODE=137
+		trap 'cleanup 2>/dev/null || rm -f ${LOCK_FILE} 2>/dev/null || true ; wait ; exit ${EXIT_CODE} ;' EXIT || EXIT_CODE=1
+else
+		# shellcheck disable=SC2046
+		printf "\t%s\n" "Check Setup Scripts Tests Coverage already in progress by "$(head "${LOCK_FILE}") >&2 ;
+		exit 126 ;
+fi
+
+# check_pip: Audit and validate Python requirements files.
+
+# THIS IS THE ACTUAL TEST DIR USED (update _TEST_ROOT_DIR as needed)
+_TEST_ROOT_DIR=$(git rev-parse --show-toplevel 2>/dev/null) ;
+if [[ -d ../.git ]] ; then
+	_TEST_ROOT_DIR="../" ;
+elif [[ -d ./.git ]] ; then
+	_TEST_ROOT_DIR=$(pwd) ;
+elif [[ ( -d $(git rev-parse --show-toplevel 2>/dev/null) ) ]] ; then
+	_TEST_ROOT_DIR=$(git rev-parse --show-toplevel 2>/dev/null) ;
+else
+	printf "\t%s\n" "FAIL: missing valid repository or source structure" >&2 ;
+	EXIT_CODE=40
+fi
+
+cd $_TEST_ROOT_DIR 2>/dev/null || EXIT_CODE=40
+
+# Get a list of files to check using git ls-tree with filtering
+FILES_TO_CHECK=$(git ls-tree -r --full-tree --name-only HEAD -- **/requirements.txt *-requirements.txt requirements.txt 2>/dev/null || EXIT_CODE=3)
+
+# Enable auto-correction if '--fix' argument is provided
+if [[ "$1" == "--fix" ]]; then
+	AUDIT_OPTIONS="--fix --strict ${AUDIT_OPTIONS}"
+	printf "%s\n" "Auto-correction enabled."
+fi
+
+# THIS IS THE ACTUAL TEST
+# Iterate over files and run checks
+for req_file in $FILES_TO_CHECK; do
+	printf "%s\n" "Checking ${req_file}" ;
+	printf "%s\n" "Auditing ${req_file} for security vulnerabilities..."
+	{ pip-audit $AUDIT_OPTIONS "${req_file}" || EXIT_CODE=4 ;} ; wait ;
+	if [[ ("${EXIT_CODE}" -eq 0) ]] ; then
+		printf "%s\n" "Checking licenses in $req_file..." ;
+		{ pip install -r "$req_file" --quiet >/dev/null || EXIT_CODE=4 ;} 2>/dev/null ; wait ;
+		{ pip-licenses $LICENSE_OPTIONS || EXIT_CODE=5 ;} 2>/dev/null ; wait ;
+	else
+		printf "%s\n" "FAIL: Found requirements errors." >&2 ;
+	fi
+done
```

Comment:
1. I see a mistake in your suggestion: if we use `$?` it will always evaluate to `true`

I respond with the trivial case.

A. consider the following bash code:
`{ false || echo $? && false ;} ; wait ; echo $? ;` which will echo out "1\n0" every time.
B. now consider the following bash code:
`{ true || echo $? && false ;} ; wait ; echo $? ;` which will echo out "0" every time.

C. Why would it always end in `0`? Unsurprisingly, because of the `wait` which invariably blocks, until it can exit with `0` and the final `echo $?` picks it up and echos it back out.

Now again, consider my approach around line 164 and focus on what the `if-else` conditional does.

2. For posterity include the official bash documentation about "$?"